File: ios/Sources/KlinikFollowUpFeature/FollowUpScreen.swift
import SwiftUI
import KlinikAPI
import KlinikCore
import KlinikDesign

/// The check-up calendar (spec M6).
public struct FollowUpScreen: View {
    @Environment(\.colorScheme) private var scheme

    private let model: FollowUpModel
    private let canMark: Bool
    /// Staff build the schedule from the operation date; a patient does not
    /// decide when they are seen.
    private let canGenerate: Bool

    @State private var state = FollowUpState()
    @State private var generating = false
    @State private var surgeryDate = Date()

    /// - Parameter canMark: staff mark a visit attended; a patient only reads.
    public init(model: FollowUpModel, canMark: Bool = false, canGenerate: Bool = false) {
        self.model = model
        self.canMark = canMark
        self.canGenerate = canGenerate
    }

    public var body: some View {
        VStack(spacing: 0) {
            content
        }
        .background(Tokens.Palette.background.resolve(for: scheme))
        .task { await refresh { await model.refresh() } }
        .refreshable { await refresh { await model.refresh() } }
        .navigationTitle(L10n.string("followUp.title"))
    }

    /**
     * Building the plan, from the one thing it depends on.
     *
     * The milestones themselves — D1, W1, M1, M3, M6, Y1 — come from the
     * server's template for the procedure. This asks for the operation date and
     * nothing else: which milestones a procedure needs is a clinical decision,
     * and a client that composed its own list would be making it.
     */
    private var buildSchedule: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: Tokens.Spacing.lg) {
                ErrorBanner(message: state.error)

                Card(tone: .info) {
                    Text(L10n.string("followUp.generateHint"))
                        .font(Tokens.Typography.bodyRelative)
                        .foregroundStyle(Tokens.Palette.textPrimary.resolve(for: scheme))
                        .fixedSize(horizontal: false, vertical: true)
                }

                DatePicker(
                    L10n.string("followUp.surgeryDate"),
                    selection: $surgeryDate,
                    displayedComponents: .date
                )
                .font(Tokens.Typography.bodyRelative)
                .frame(minHeight: Tokens.minimumTouchTarget)

                PrimaryButton(
                    title: L10n.string("followUp.generate"),
                    isBusy: generating,
                    isEnabled: !generating
                ) {
                    generating = true
                    _ = await model.generate(surgeryDate: surgeryDate)
                    state = await model.currentState()
                    generating = false
                }
            }
            .padding(Tokens.Spacing.lg)
        }
    }

    @ViewBuilder
    private var content: some View {
        switch state.phase {
        case .loading:
            Spacer()
            ProgressView().accessibilityLabel(L10n.string("common.loading"))
            Spacer()

        case .none:
            if canGenerate {
                buildSchedule
            } else {
                MessageState(icon: "calendar", text: L10n.string("followUp.empty"))
            }

        case .notFound:
            MessageState(icon: "questionmark.folder", text: L10n.string("error.notFound"))

        case .failed(let message):
            MessageState(
                icon: Tokens.State.labCritical.iconName,
                text: message,
                retryTitle: L10n.string("common.retry")
            ) {
                await refresh { await model.refresh() }
            }

        case .loaded:
            schedule
        }
    }

    private var schedule: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: Tokens.Spacing.lg) {
                // The next visit, first and large. It is what a patient opens
                // this screen to find, and a list they have to read to the
                // middle of is one they read wrong.
                if let next = state.next {
                    NextVisitCard(milestone: next)
                }

                if !state.missed.isEmpty {
                    Label(
                        "\(L10n.string("followUp.missedCount")): \(state.missed.count)",
                        systemImage: Tokens.State.triageUrgent.iconName
                    )
                    .font(Tokens.Typography.calloutRelative)
                    .foregroundStyle(Tokens.Palette.warning.resolve(for: scheme))
                }

                if let error = state.error {
                    ErrorBanner(message: error)
                }

                ForEach(state.schedule?.milestones ?? []) { milestone in
                    MilestoneRow(
                        milestone: milestone,
                        canMark: canMark,
                        isWorking: state.working == milestone.id
                    ) { status in
                        await refresh { await model.mark(milestone.id, as: status) }
                    }
                }
            }
            .padding(Tokens.Spacing.lg)
        }
    }

    private func refresh(_ work: () async -> Void) async {
        await work()
        state = await model.currentState()
    }
}

struct NextVisitCard: View {
    @Environment(\.colorScheme) private var scheme

    let milestone: Milestone

    var body: some View {
        VStack(alignment: .leading, spacing: Tokens.Spacing.xs) {
            Text(L10n.string("followUp.nextVisit"))
                .font(Tokens.Typography.captionRelative)
                .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))

            Text(milestone.localizedLabel)
                .font(Tokens.Typography.headingRelative)
                .foregroundStyle(Tokens.Palette.textPrimary.resolve(for: scheme))

            Text(milestone.dueAt, style: .date)
                .font(Tokens.Typography.bodyRelative)
                .foregroundStyle(Tokens.Palette.textPrimary.resolve(for: scheme))
        }
        .padding(Tokens.Spacing.lg)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Tokens.Palette.infoSurface.resolve(for: scheme))
        .clipShape(RoundedRectangle(cornerRadius: Tokens.Radius.md))
        .accessibilityElement(children: .combine)
    }
}

struct MilestoneRow: View {
    @Environment(\.colorScheme) private var scheme

    let milestone: Milestone
    let canMark: Bool
    let isWorking: Bool
    let onMark: (MilestoneStatus) async -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: Tokens.Spacing.xs) {
            HStack {
                Text(milestone.localizedLabel)
                    .font(Tokens.Typography.subheadingRelative)
                    .foregroundStyle(Tokens.Palette.textPrimary.resolve(for: scheme))

                Spacer()

                // State in words as well as colour, because a colour a reader
                // cannot distinguish says nothing (spec section 7).
                Text(milestone.status.localizedName)
                    .font(Tokens.Typography.captionRelative)
                    .foregroundStyle(tint.resolve(for: scheme))
            }

            Text(milestone.dueAt, style: .date)
                .font(Tokens.Typography.captionRelative)
                .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))

            if canMark && milestone.status.isOutstanding {
                HStack(spacing: Tokens.Spacing.md) {
                    Button(L10n.string("followUp.markAttended")) {
                        Task { await onMark(.completed) }
                    }
                    .disabled(isWorking)
                    .frame(minHeight: Tokens.minimumTouchTarget)

                    Button(L10n.string("followUp.markSkipped")) {
                        Task { await onMark(.skipped) }
                    }
                    .disabled(isWorking)
                    .frame(minHeight: Tokens.minimumTouchTarget)
                }
            }
        }
        .padding(.vertical, Tokens.Spacing.xs)
        .accessibilityElement(children: .combine)
    }

    private var tint: ThemedColor {
        switch milestone.status {
        case .completed: return Tokens.Palette.success
        case .missed: return Tokens.Palette.warning
        case .skipped: return Tokens.Palette.textSecondary
        case .pending, .notified: return Tokens.Palette.info
        }
    }
}

File: android/feature/followup-ui/src/main/kotlin/xyz/klinik/feature/followup/ui/FollowUpScreen.kt
package xyz.klinik.feature.followup.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import xyz.klinik.design.Tokens
import xyz.klinik.design.klinikColor
import xyz.klinik.feature.followup.FollowUpPhase
import xyz.klinik.feature.followup.FollowUpState
import xyz.klinik.network.Milestone
import xyz.klinik.network.MilestoneStatus
import xyz.klinik.network.UiText

data class FollowUpStrings(
    val empty: String,
    val notFound: String,
    val retry: String,
    val nextVisit: String,
    val missedCount: String,
    val markAttended: String,
    val markSkipped: String,
    val milestoneName: (Milestone) -> String,
    val statusName: (MilestoneStatus) -> String,
    val message: (UiText) -> String,
)

/** The check-up calendar generated from the operation date (spec M6). */
@Composable
fun FollowUpScreen(
    state: FollowUpState,
    strings: FollowUpStrings,
    /** Staff mark a visit attended; a patient only reads. */
    canMark: Boolean,
    /**
     * Now, as ISO-8601, supplied by the caller.
     *
     * A screen that reads the clock itself cannot be tested at the one boundary
     * that matters here — which visit counts as next.
     */
    nowIso: String,
    onRetry: () -> Unit,
    onMark: (milestoneId: String, status: MilestoneStatus) -> Unit,
    modifier: Modifier = Modifier,
) {
    Surface(color = klinikColor("background"), modifier = modifier.fillMaxSize()) {
        when (val phase = state.phase) {
            FollowUpPhase.Loading -> Centered { CircularProgressIndicator() }

            FollowUpPhase.None -> Centered { Text(strings.empty) }

            FollowUpPhase.NotFound -> Centered { Text(strings.notFound) }

            is FollowUpPhase.Failed -> Centered {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Text(strings.message(UiText.Key(phase.messageKey)))
                    TextButton(onClick = onRetry) { Text(strings.retry) }
                }
            }

            FollowUpPhase.Loaded -> Column(
                modifier = Modifier
                    .fillMaxSize()
                    .verticalScroll(rememberScrollState())
                    .padding(Tokens.Spacing.lg),
                verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.lg),
            ) {
                // The next visit, first and large. It is what a patient opens
                // this screen to find, and a list they have to read to the
                // middle of is one they read wrong.
                state.next(nowIso)?.let { next ->
                    Surface(color = klinikColor("infoSurface"), modifier = Modifier.fillMaxWidth()) {
                        Column(
                            modifier = Modifier.padding(Tokens.Spacing.lg),
                            verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.xs),
                        ) {
                            Text(strings.nextVisit, color = klinikColor("textSecondary"))
                            Text(
                                strings.milestoneName(next),
                                color = klinikColor("textPrimary"),
                                modifier = Modifier.semantics { heading() },
                            )
                            Text(next.dueAt, color = klinikColor("textPrimary"))
                        }
                    }
                }

                if (state.missed.isNotEmpty()) {
                    Text(
                        "${strings.missedCount}: ${state.missed.size}",
                        color = klinikColor("warning"),
                    )
                }

                state.error?.let { Text(strings.message(it), color = klinikColor("critical")) }

                for (milestone in state.schedule?.milestones.orEmpty()) {
                    MilestoneRow(milestone, strings, canMark, state.working == milestone.id, onMark)
                }
            }
        }
    }
}

@Composable
private fun MilestoneRow(
    milestone: Milestone,
    strings: FollowUpStrings,
    canMark: Boolean,
    isWorking: Boolean,
    onMark: (String, MilestoneStatus) -> Unit,
) {
    Column(verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.xs)) {
        Row(modifier = Modifier.fillMaxWidth()) {
            Text(
                strings.milestoneName(milestone),
                color = klinikColor("textPrimary"),
                modifier = Modifier.weight(1f),
            )

            // State in words as well as colour, because a colour a reader
            // cannot distinguish says nothing (spec section 7).
            Text(strings.statusName(milestone.status), color = tintFor(milestone.status))
        }

        Text(milestone.dueAt, color = klinikColor("textSecondary"))

        if (canMark && milestone.status.isOutstanding) {
            Row(horizontalArrangement = Arrangement.spacedBy(Tokens.Spacing.md)) {
                TextButton(
                    onClick = { onMark(milestone.id, MilestoneStatus.COMPLETED) },
                    enabled = !isWorking,
                    modifier = Modifier.heightIn(min = Tokens.minimumTouchTarget),
                ) { Text(strings.markAttended) }

                TextButton(
                    onClick = { onMark(milestone.id, MilestoneStatus.SKIPPED) },
                    enabled = !isWorking,
                    modifier = Modifier.heightIn(min = Tokens.minimumTouchTarget),
                ) { Text(strings.markSkipped) }
            }
        }
    }
}

@Composable
private fun tintFor(status: MilestoneStatus) = when (status) {
    MilestoneStatus.COMPLETED -> klinikColor("success")
    MilestoneStatus.MISSED -> klinikColor("warning")
    MilestoneStatus.SKIPPED -> klinikColor("textSecondary")
    MilestoneStatus.PENDING, MilestoneStatus.NOTIFIED -> klinikColor("info")
}

@Composable
private fun Centered(content: @Composable () -> Unit) {
    Box(
        modifier = Modifier.fillMaxSize().padding(Tokens.Spacing.xl),
        contentAlignment = Alignment.Center,
    ) { content() }
}
